Fix default network drivers

bazsi opened the pull request:
This should fix #1945 

@lbudai @MrAnno please review the SSL piece, as it might somewhat be questionable, but I am open for suggestions.

kira-syslogng commented:
success

MrAnno commented on lib/tlscontext.c:
```diff
@@ -602,15 +602,15 @@ _are_key_and_cert_files_accessible(TLSContext *self)
-_client_key_and_cert_files_are_not_specified(TLSContext *self)
+_key_and_cert_files_are_not_specified(TLSContext *self)
 {
-  return self->mode == TM_CLIENT && (!self->key_file && !self->cert_file);
```
Omitting the key in server mode should not be allowed in my opinion.
There is no secure cipher in OpenSSL that works without a key, so every connection attempt will fail with an SSL error: "no shared cipher".

I understand that this is required to fix `default-network-drivers` (since we can't conditionally remove a source from the config based on the value of a SCL block parameter). Adding such a feature to the grammar would be a much bigger job, so I can reconcile with this patch :smile: .

MrAnno commented:
Omitting the key in server mode should not be allowed in my opinion.
There is no secure cipher in OpenSSL that works without a key, so every connection attempt will fail with an SSL error: "no shared cipher".

I understand that this is required to fix `default-network-drivers` (since we can't conditionally remove a source from the config based on the value of a SCL block parameter). Adding such a feature to the grammar would be a much bigger job, so I can reconcile with this patch :smile: .

MrAnno commented:
Just a question:
Why don't we remove the TLS source from `default-network-drivers`? For me, default means "it can work without additional configuration". I know it's a weird question, but I'm interested in your opinion about this.

gaborznagy commented:
@bazsi, I've put the WIP flag on the PR according to the today's discussion (adding some warning messages, but correct me if I don't remember correctly).
You have also explained why should the tls transport included in `default-network-drivers`, i.e. you  would not prefer to have separate `default-network-drivers` for TLS. And also why should the empty `tls()` block accepted in the config.

mitzkia commented:
@kira-syslogng test this please

kira-syslogng commented:
success

bazsi commented:
I have now rebased this against master and also addressed all the comments.

It also improves the TLS related error handling so that error messages include the tls() location in the configuration file.

I've removed the WIP mark, so once it is green  it should be ready to go in and fix #1945

kira-syslogng commented:
success

kira-syslogng commented:
success

MrAnno commented:
Looks really good. 👍

furiel commented on lib/tlscontext.c:
```diff
@@ -321,30 +328,39 @@ tls_session_free(TLSSession *self)
+      msg_error("Error opening TLS related key or certificate file",
                 evt_tag_str("filename", fname),
-                evt_tag_errno("error", errno));
+                evt_tag_errno("error", errno),
```
This is a general problem, but it is not safe to use errno this way. The malloc, strdup called from evt_tag_str can overwrite it. As the parameters are evaluated in any order, it is not even enough to move it to the first parameter.

As the original code had this defect too, and probably we will not run into this from practical point of view, I leave just a note here, and I create an issue to fix these in one swipe within the code. Maybe we can even assert if errno is passed directly to evt_tag_errno.

furiel commented:
@bazsi @fekete-robert as I see default-network-sources is part of the code since 3.12, but I do not see it in the documentation. Should we write something about it?
